File: src/main/java/me/meta4245/betterthanmodern/mixin/ChickenEntityMixin.java
package me.meta4245.betterthanmodern.mixin;

import me.meta4245.betterthanmodern.event.ItemRegistry;
import me.meta4245.betterthanmodern.mixin.accessor.EntityAccessor;
import net.minecraft.entity.passive.ChickenEntity;
import net.minecraft.item.Item;
import net.minecraft.world.World;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Unique;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.util.List;
import java.util.Random;

@Mixin(ChickenEntity.class)
public abstract class ChickenEntityMixin extends AnimalEntityMixin {
    @Unique
    private final static List<Item> breedFood = List.of(
            ItemRegistry.melonSeeds,
            Item.SEEDS
    );

    public ChickenEntityMixin(World world) {
        super(world);
    }

    @Inject(
            method = "getDroppedItemId",
            at = @At("HEAD"),
            cancellable = true
    )
    private void getDroppedItemId(CallbackInfoReturnable<Integer> cir) {
        EntityAccessor accessor = (EntityAccessor) this;
        Random random = accessor.getRandom();

        int featherAmount = random.nextInt(0, 3);
        int id = accessor.getFireTicks() > 0
                ? ItemRegistry.cookedChicken.id
                : ItemRegistry.rawChicken.id;

        accessor.callDropItem(Item.FEATHER.id, featherAmount);
        accessor.callDropItem(id, 1);

        // LivingEntity doesn't drop if getDroppedId returns 0
        cir.setReturnValue(0);
    }

    @Override
    protected List<Item> getBreedFood() {
        return breedFood;
    }
}

File: src/main/kotlin/me/meta4245/betterthanmodern/mixin/ChickenEntityMixin.kt
package me.meta4245.betterthanmodern.mixin

import me.meta4245.betterthanmodern.mixin.accessor.EntityAccessor
import net.minecraft.entity.passive.ChickenEntity
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(ChickenEntity::class)
abstract class ChickenEntityMixin {
    @Inject(at = [At("HEAD")], method = ["getDroppedId"], cancellable = true)
    fun getDroppedId(cir: CallbackInfoReturnable<Int?>?) {
        val accessor = this as EntityAccessor
        val fireTicks = accessor.fireTicks

        //        int rawChickenId;
//        int cookedChickenId;
//
//        try {
//            rawChickenId = (int) ItemRegistry.class.getDeclaredField("RawChickenId").get(this);
//            cookedChickenId = (int) ItemRegistry.class.getDeclaredField("CookedChickenId").get(this);
//        } catch (Exception e) {
//            throw new RuntimeException(e);
//        }
//
//        cir.setReturnValue(
//                fireTicks > 0
//                        ? cookedChickenId
//                        : rawChickenId
//        );
    }
}
